internal/tests/api/credentials/credentials_test.go

```diff
@@ -384,3 +394,112 @@ func TestErrors(t *testing.T) {
+	hcClient := hostcatalogs.NewClient(client)
+	_, proj := iam.TestScopes(t, tc.IamRepo(), iam.WithUserId(token.UserId))
+
+	hc, err := hcClient.Create(ctx, "static", proj.PublicId, hostcatalogs.WithName("my-host-catalog"))
+	require.NoError(err)
+	host, err := hostClient.Create(ctx, hc.Item.Id, hosts.WithName("my-host"), hosts.WithStaticHostAddress("example.com"))
+	require.NoError(err)
+	hs, err := hsClient.Create(ctx, hc.Item.Id, hostsets.WithName("my-host-set"))
+	require.NoError(err)
+	_, err = hsClient.AddHosts(ctx, hs.Item.Id, 1, []string{host.Item.Id})
+	require.NoError(err)
+	cs, err := credentialstores.NewClient(client).Create(tc.Context(), "static", proj.GetPublicId())
+	require.NoError(err)
+	obj := map[string]interface{}{
+		"username": "admin",
+		"password": "pass",
+	}
+	cred, err := credsClient.Create(ctx, credential.JsonSubtype.String(), cs.Item.Id, credentials.WithName("foo"), credentials.WithJsonCredentialObject(obj))
+	require.NoError(err)
+	targ, err := tgClient.Create(ctx, "tcp", proj.PublicId, targets.WithName("my-target"), targets.WithTcpTargetDefaultPort(22))
+	require.NoError(err)
+	_, err = tgClient.AddHostSources(ctx, targ.Item.Id, 1, []string{hs.Item.Id})
+	require.NoError(err)
+	_, err = tgClient.AddCredentialSources(ctx, targ.Item.Id, 2, targets.WithBrokeredCredentialSourceIds([]string{cred.Item.Id}))
+	w := worker.NewTestWorker(t, &worker.TestWorkerOpts{
+		InitialUpstreams: tc.ClusterAddrs(),
+		Logger:           logger.Named("worker"),
+		WorkerAuthKms:    tc.Config().WorkerAuthKms,
+		Name:             "worker",
+	})
+	t.Cleanup(w.Shutdown)
+	require.NoError(w.Worker().WaitForNextSuccessfulStatusUpdate())
+
+	// Authorize session, requires decrypting json credential
+	_, err = tgClient.AuthorizeSession(ctx, targ.Item.Id)
+	require.NoError(err)
+
+	// Create new key versions
+	_, err = scopesClient.RotateKeys(ctx, proj.PublicId, false)
+	require.NoError(err)
+
+	// Update JSON credential, will re-encrypt with new key versions
+	obj["password"] = "password"
+	cred, err = credsClient.Update(ctx, cred.Item.Id, 1, credentials.WithJsonCredentialObject(obj))
+	require.NoError(err)
+
+	// Create new key versions again
+	_, err = scopesClient.RotateKeys(ctx, proj.PublicId, false)
+	require.NoError(err)
+
+	keys, err := scopesClient.ListKeys(ctx, proj.PublicId)
+	require.NoError(err)
+
+	var destroyKeyVersion *scopes.KeyVersion
+
+	for _, key := range keys.Items {
+		if key.Purpose == kms.KeyPurposeDatabase.String() {
+			// Versions are sorted in descending order, this gets the 2nd version.
+			// It is the keyversion that was used to re-encrypt our JSON credential.
+			destroyKeyVersion = key.Versions[1]
+			break
+		}
+	}
+
+	// Destroy the older key version
+	result, err := scopesClient.DestroyKeyVersion(ctx, proj.PublicId, destroyKeyVersion.Id)
+	require.NoError(err)
+	// Should start asynchronous rewrapping of the encrypted JSON credential
+	assert.Equal("pending", result.State)
```

Comment:
what is the purpose of destroying the key version here? could we not just pull this row from the db and ensure that the key id has been updated? perhaps attempt decrypting with that new keyid?

Reply:
The reason we're destroying the key version here is because I wanted to create a test case that could realistically be reproduced by a user. This is an integration test, the test oracle for this isn't "is the DB row updated?" it's "does the product behave correctly under this specific scenario using its public API endpoints?". This way, this test doesn't have to know anything about our internal implementation details, just like the user doesn't know about it.